commit 1edce7165e890355f1b5e27f863541af70a1d20d (refs/changes/03/1403/6)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-02-14T16:06:29+01:00 (2 years, 8 months ago)
    
    TOOLS-1645 sdcadm update procedure for HA-ready stateless services
    Reviewed by: Trent Mick <trent.mick@joyent.com>

diff --git a/lib/procedures/index.js b/lib/procedures/index.js
index 9e5c3b5..7c7ca55 100644
--- a/lib/procedures/index.js
+++ b/lib/procedures/index.js
@@ -48,8 +48,8 @@ NoOp.prototype.execute = function noOpExecute(options, cb) {
 
 // --- procedures from the modules where we've defined them:
 var DownloadImages = require('./download-images').DownloadImages;
-var UpdateStatelessServicesV1 =
-    require('./update-stateless-services-v1').UpdateStatelessServicesV1;
+var UpdateStatelessServices =
+    require('./update-stateless-services-v1').UpdateStatelessServices;
 var UpdateSingleHeadnodeImgapi =
     require('./update-single-headnode-imgapi').UpdateSingleHeadnodeImgapi;
 var UpdateUFDSServiceV1 =
@@ -59,7 +59,6 @@ var UpdateSingleHNSapiV1 =
     require('./update-single-hn-sapi-v1').UpdateSingleHNSapiV1;
 var UpdateManateeV2 = require('./update-manatee-v2').UpdateManateeV2;
 var UpdateBinderV2 = require('./update-binder-v2').UpdateBinderV2;
-var UpdateMahiV2 = require('./update-mahi-v2').UpdateMahiV2;
 // --- Create service instance also from procedures:
 var CreateServiceInstanceV1 =
 require('./create-service-instance-v1').CreateServiceInstanceV1;
@@ -160,25 +159,17 @@ function coordinatePlan(opts, cb) {
         },
 
         /**
-         * Update services that are (a) stateless, (b) have a single instance
-         * **on the headnode**, (c) with no current special handling (like
-         * migrations).
-         *
-         * Here (b) implies this is the early SDC world where we don't have
-         * HA multiple instances of services.
+         * Update services that are stateless with no current special handling
+         * (like migrations).
          */
         function updateSimpleServices(_, next) {
-            var simpleServices = ['adminui', 'amon', 'amonredis', 'assets',
-                'ca', 'cloudapi', 'cnapi', 'dhcpd', 'docker', 'fwapi',
-                'cmon',
-                'cns',
-                'manta',
-                'napi', 'portolan',
-                'papi',
-                'rabbitmq', 'redis', 'sdc', 'vmapi', 'workflow'];
+            var simpleServices = [
+                'adminui', 'amon', 'amonredis', 'assets', 'ca',
+                'cnapi', 'cns', 'dhcpd', 'docker', 'fwapi', 'manta',
+                'napi', 'rabbitmq', 'redis', 'sdc', 'vmapi'
+            ].concat(HA_READY_SVCS);
             var handle = [];
             var remaining = [];
-            var currHostname = os.hostname();
             changes.forEach(function (change) {
                 var svcInsts = instsFromSvcName[change.service.name] || [];
                 if (change.type === 'update-instance' &&
@@ -194,7 +185,7 @@ function coordinatePlan(opts, cb) {
                         log.debug({
                                 numInsts: 0,
                                 svc: change.service.name
-                            }, 'UpdateStatelessServicesV1 update service ' +
+                            }, 'UpdateStatelessServices update service ' +
                             'with no instance');
 
                         if (!opts.noVerbose) {
@@ -209,18 +200,17 @@ function coordinatePlan(opts, cb) {
                             change.insts = svcInsts;
                             handle.push(change);
                         } else {
-                            log.debug({
-                                    numInsts: svcInsts.length,
-                                    svc: change.service.name
-                                }, 'UpdateStatelessServicesV1 skip change: ' +
-                                'not 1 inst');
+                            if (~HA_READY_SVCS.indexOf(change.service.name)) {
+                                change.insts = svcInsts;
+                                handle.push(change);
+                            } else {
+                                log.debug({
+                                        numInsts: svcInsts.length,
+                                        svc: change.service.name
+                                    }, 'UpdateStatelessServices skip change: ' +
+                                    'not 1 inst');
+                            }
                         }
-                    } else if (svcInsts[0].hostname !== currHostname) {
-                        log.debug({
-                                svc: change.service.name,
-                                cn: svcInsts[0].server
-                            }, 'UpdateStatelessServicesV1 skip change: ' +
-                            'inst not on headnode');
                     } else {
                         change.inst = svcInsts[0];
                         handle.push(change);
@@ -232,9 +222,9 @@ function coordinatePlan(opts, cb) {
             if (handle.length) {
                 changes = remaining;
                 log.debug({changes: handle},
-                    'UpdateStatelessServicesV1 will handle %d change(s)',
+                    'UpdateStatelessServices will handle %d change(s)',
                     handle.length);
-                procs.push(new UpdateStatelessServicesV1({
+                procs.push(new UpdateStatelessServices({
                     changes: handle
                 }));
             }
@@ -570,40 +560,6 @@ function coordinatePlan(opts, cb) {
             next();
         },
 
-        function updateMahi(_, next) {
-            var handle = [];
-            var remaining = [];
-            changes.forEach(function (change) {
-                if (change.type === 'update-instance' &&
-                    change.service.name === 'mahi')
-                {
-                    change.inst = change.instance;
-                    handle.push(change);
-                } else if ((change.type === 'update-service' ||
-                    change.type === 'rollback-service') &&
-                    change.service.name === 'mahi')
-                {
-                    var svcInsts = instsFromSvcName[change.service.name] || [];
-                    if (svcInsts.length && svcInsts.length > 1) {
-                        change.insts = svcInsts;
-                    } else {
-                        change.inst = svcInsts[0];
-                    }
-                    handle.push(change);
-                } else {
-                    remaining.push(change);
-                }
-            });
-            if (handle.length) {
-                changes = remaining;
-                log.debug({changes: handle},
-                    'UpdateMahiV2 will handle %d change(s)',
-                    handle.length);
-                procs.push(new UpdateMahiV2({ changes: handle }));
-            }
-            next();
-        },
-
 
         /**
          * Create simple service instances. It's to say, do not create those
diff --git a/lib/procedures/update-mahi-v2.js b/lib/procedures/update-mahi-v2.js
deleted file mode 100644
index f203ca5..0000000
--- a/lib/procedures/update-mahi-v2.js
+++ /dev/null
@@ -1,209 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright 2016 Joyent, Inc.
- */
-
-var assert = require('assert-plus');
-var sprintf = require('extsprintf').sprintf;
-var util = require('util');
-var vasync = require('vasync');
-
-var common = require('../common');
-var errors = require('../errors');
-
-var Procedure = require('./procedure').Procedure;
-var s = require('./shared');
-
-/**
- * Procedure for updating mahi service, HA.
- */
-function UpdateMahiV2(options) {
-    assert.arrayOfObject(options.changes, 'options.changes');
-    this.changes = options.changes;
-}
-util.inherits(UpdateMahiV2, Procedure);
-
-UpdateMahiV2.prototype.summarize = function ushiSummarize() {
-    if (this.changes[0].type === 'update-instance') {
-        return this.changes.map(function (ch) {
-            return sprintf('update instance "%s" (%s)\n' +
-                    'of service "%s" to image %s\n', ch.inst.instance,
-                    ch.inst.alias, ch.service.name, ch.image.uuid) +
-                common.indent(sprintf('(%s@%s)',
-                    ch.image.name, ch.image.version));
-        }).join('\n');
-    } else {
-        var word = (this.changes[0].type === 'rollback-service') ?
-            'rollback' : 'update';
-        var c0 = this.changes[0];
-        var img = c0.image;
-        var out = [sprintf('%s "%s" service to image %s', word,
-                        c0.service.name, img.uuid),
-                    common.indent(sprintf('(%s@%s)', img.name, img.version))];
-        if (c0.insts) {
-            out[0] += ':';
-            out = out.concat(c0.insts.map(function (inst) {
-                return common.indent(sprintf('instance "%s" (%s) in server %s',
-                    inst.zonename, inst.alias, inst.server));
-            }));
-        }
-        return out.join('\n');
-    }
-};
-
-UpdateMahiV2.prototype.execute = function ushiExecute(opts, cb) {
-    assert.object(opts, 'opts');
-    assert.object(opts.sdcadm, 'opts.sdcadm');
-    assert.object(opts.plan, 'opts.plan');
-    assert.object(opts.log, 'opts.log');
-    assert.func(opts.progress, 'opts.progress');
-    assert.string(opts.wrkDir, 'opts.wrkDir');
-    assert.func(cb, 'cb');
-    var self = this;
-    var progress = opts.progress;
-    var rollback = opts.plan.rollback || false;
-
-    function updateMahi(change, nextSvc) {
-        var inst = change.inst;
-
-        var arg = {
-            change: change,
-            opts: opts,
-            userScript: false,
-            HA: false,
-            tmpAlias: null,
-            tmpUUID: null
-        };
-
-        if ((change.insts && change.insts.length > 1) || change.HA) {
-            arg.HA = true;
-        } else {
-            arg.tmpAlias = inst.alias + 'tmp';
-        }
-
-        if (opts.plan.changes.length > 1) {
-            progress('');
-            progress('--- Updating %s ...', change.service.name);
-        }
-
-        var funcs = [];
-
-        if (rollback) {
-            funcs.push(s.getOldUserScript);
-        } else {
-            funcs.push(s.getUserScript);
-            funcs.push(s.writeOldUserScriptForRollback);
-        }
-
-        funcs.push(s.updateSvcUserScript);
-
-        if (arg.HA) {
-            change.insts.forEach(function (ins) {
-                funcs.push(function (_, next) {
-                    s.ensureDelegateDataset({
-                        service: change.service,
-                        progress: progress,
-                        zonename: ins.zonename,
-                        log: opts.log,
-                        server: ins.server
-                    }, next);
-                });
-                funcs.push(function (_, next) {
-                    s.updateVmUserScriptRemote({
-                        service: change.service,
-                        progress: progress,
-                        zonename: ins.zonename,
-                        log: opts.log,
-                        server: ins.server,
-                        userScript: arg.userScript
-                    }, next);
-                });
-            });
-        } else {
-            funcs.push(function (_, next) {
-                s.ensureDelegateDataset({
-                    service: change.service,
-                    progress: progress,
-                    zonename: inst.zonename,
-                    log: opts.log,
-                    server: inst.server
-                }, next);
-            });
-            funcs.push(s.updateVmUserScript);
-        }
-
-        funcs.push(s.updateSapiSvc);
-
-        if (arg.HA) {
-            change.insts.forEach(function (ins) {
-                funcs = funcs.concat(
-                    function imgadmInstall(_, next) {
-                        return s.imgadmInstallRemote({
-                            progress: progress,
-                            img: change.image,
-                            log: opts.log,
-                            server: ins.server
-                        }, next);
-                    },
-                    function reprovisionInst(_, next) {
-                        s.reprovisionRemote({
-                            server: ins.server,
-                            img: change.image,
-                            zonename: ins.zonename,
-                            progress: progress,
-                            log: opts.log,
-                            sdcadm: opts.sdcadm
-                        }, next);
-                    },
-                    function waitForInstToBeUp(_, next) {
-                        progress('Wait (sleep) for %s instance %s to come up',
-                            ins.service, ins.zonename);
-                        setTimeout(next, 15 * 1000);
-                    }
-                );
-            });
-        } else {
-            funcs = funcs.concat([
-                s.imgadmInstall,
-                s.reprovision,
-                s.waitForInstToBeUp
-            ]);
-        }
-        vasync.pipeline({funcs: funcs, arg: arg}, nextSvc);
-    }
-
-    // TOOLS-1465: when updating individual instances, need to double check if
-    // service is part or not of an HA setup
-    function checkServiceHA(change, nextSvc) {
-        opts.sdcadm.sapi.listInstances({
-            service_uuid: change.service.uuid
-        }, function (err, insts) {
-            if (err) {
-                return nextSvc(new errors.SDCClientError(err, 'SAPI'));
-            }
-            if (insts.length > 1) {
-                change.HA = true;
-                if (!change.insts) {
-                    change.insts = [change.inst];
-                }
-            }
-            return updateMahi(change, nextSvc);
-        });
-    }
-
-    vasync.forEachPipeline({
-        inputs: self.changes,
-        func: checkServiceHA
-    }, cb);
-};
-//---- exports
-
-module.exports = {
-    UpdateMahiV2: UpdateMahiV2
-};
-// vim: set softtabstop=4 shiftwidth=4:
diff --git a/lib/procedures/update-stateless-services-v1.js b/lib/procedures/update-stateless-services-v1.js
index 2720e56..a837e82 100644
--- a/lib/procedures/update-stateless-services-v1.js
+++ b/lib/procedures/update-stateless-services-v1.js
@@ -5,57 +5,75 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var assert = require('assert-plus');
 var sprintf = require('extsprintf').sprintf;
 var util = require('util');
-
 var vasync = require('vasync');
 
 var Procedure = require('./procedure').Procedure;
 var s = require('./shared');
 var common = require('../common');
+var errors = require('../errors');
 
 /**
- * This is a limited first pass procedure for updating a set of stateless SDC
- * services.
+ * Procedure for updating a set of stateless SDC services.
+ *
+ * This procedure doesn't have the previous version limitations of only being
+ * able to update single service instances constrained to the Headnode.
  *
  * Limitations:
- * - the service must only have one instance
- * - the instance must be on the headnode (where `sdcadm` is running)
  * - we only support the "stateless" easy-to-update services that don't require
  *   any migrations, bootstrapping, etc.
  */
-function UpdateStatelessServicesV1(options) {
+function UpdateStatelessServices(options) {
     assert.arrayOfObject(options.changes, 'options.changes');
     this.changes = options.changes;
 }
-util.inherits(UpdateStatelessServicesV1, Procedure);
-
-UpdateStatelessServicesV1.prototype.summarize = function ussv1Summarize() {
-    if (this.changes[0].type === 'update-instance') {
-        return this.changes.map(function (ch) {
-            return sprintf('update instance "%s" (%s)\n' +
-                    'of service "%s" to image %s\n', ch.inst.instance,
-                    ch.inst.alias, ch.service.name, ch.image.uuid) +
-                common.indent(sprintf('(%s@%s)',
-                    ch.image.name, ch.image.version));
-        }).join('\n');
-    } else {
-        var word = (this.changes[0].type === 'rollback-service') ?
-            'rollback' : 'update';
-        return this.changes.map(function (ch) {
-            return sprintf('%s "%s" service to image %s\n', word,
-                    ch.service.name, ch.image.uuid) +
-                common.indent(sprintf('(%s@%s)',
-                    ch.image.name, ch.image.version));
-        }).join('\n');
-    }
+util.inherits(UpdateStatelessServices, Procedure);
+
+
+UpdateStatelessServices.prototype.summarize = function ussv2Summarize() {
+    var out = [];
+    this.changes.forEach(function (ch) {
+        if (ch.type === 'update-instance') {
+            out.push(sprintf('update instance "%s" (%s)\n' +
+                        'of service "%s" to image %s\n', ch.inst.instance,
+                        ch.inst.alias, ch.service.name, ch.image.uuid),
+                    common.indent(sprintf('(%s@%s)',
+                        ch.image.name, ch.image.version)));
+        } else {
+            var word = (ch.type === 'rollback-service') ?
+                'rollback' : 'update';
+            var img = ch.image;
+            var msg = sprintf('%s "%s" service to image %s\n',
+                        word, ch.service.name, img.uuid) +
+                    common.indent(sprintf('(%s@%s)', img.name, img.version));
+
+            if (ch.insts) {
+                msg += ':\n';
+                msg += ch.insts.map(function (inst) {
+                    return common.indent(sprintf(
+                        'instance "%s" (%s) on server %s',
+                        inst.zonename, inst.alias, inst.server));
+                }).join('\n');
+            } else if (ch.inst) {
+                msg += ':\n';
+                msg += common.indent(sprintf(
+                        'instance "%s" (%s) on server %s',
+                        ch.inst.zonename, ch.inst.alias, ch.inst.server));
+            }
+            out.push(msg);
+        }
+    });
+
+    return out.join('\n');
+
 };
 
-UpdateStatelessServicesV1.prototype.execute = function ussv1Execute(opts, cb) {
+UpdateStatelessServices.prototype.execute = function ussv2Execute(opts, cb) {
     assert.object(opts, 'opts');
     assert.object(opts.sdcadm, 'opts.sdcadm');
     assert.object(opts.plan, 'opts.plan');
@@ -63,41 +81,68 @@ UpdateStatelessServicesV1.prototype.execute = function ussv1Execute(opts, cb) {
     assert.func(opts.progress, 'opts.progress');
     assert.string(opts.wrkDir, 'opts.wrkDir');
     assert.func(cb, 'cb');
+
     var self = this;
+    var progress = opts.progress;
     var rollback = opts.plan.rollback || false;
 
-    function updateSvc(change, nextSvc) {
+    function updateService(change, nextSvc) {
+        var inst = change.inst;
+
         var arg = {
             change: change,
             opts: opts,
-            userScript: false
+            userScript: false,
+            HA: false,
+            tmpUUID: null
         };
 
-        if (opts.plan.changes.length > 1) {
-            opts.progress('');
-            opts.progress('--- Updating %s ...', change.service.name);
+        if ((change.insts && change.insts.length > 1) || change.HA) {
+            arg.HA = true;
         }
 
-        var steps = [];
-        if (rollback) {
-            steps.push(s.getOldUserScript);
-        } else {
-            steps.push(s.getUserScript);
-            steps.push(s.writeOldUserScriptForRollback);
+        if (opts.plan.changes.length > 1) {
+            progress('');
+            progress('--- Updating %s ...', change.service.name);
         }
+
+        var funcs = [];
         if (change.service.metadata) {  // workaround for assets (TOOLS-695)
             if (rollback) {
-                steps.push(s.getOldUserScript);
+                funcs.push(s.getOldUserScript);
             } else {
-                steps.push(s.getUserScript);
-                steps.push(s.writeOldUserScriptForRollback);
+                funcs.push(s.getUserScript);
+                funcs.push(s.writeOldUserScriptForRollback);
             }
-            steps = steps.concat([
-                s.updateSvcUserScript,
-                s.updateSapiSvc
-            ]);
+
+            funcs.push(s.updateSvcUserScript);
         }
-        if (change.inst) {
+
+        if (arg.HA) { // Assume assets will not be HA setup for now
+            change.insts.forEach(function (ins) {
+                if (change.service.params.delegate_dataset) {
+                    funcs.push(function (_, next) {
+                        s.ensureDelegateDataset({
+                            service: change.service,
+                            progress: progress,
+                            zonename: ins.zonename,
+                            log: opts.log,
+                            server: ins.server
+                        }, next);
+                    });
+                }
+                funcs.push(function (_, next) {
+                    s.updateVmUserScriptRemote({
+                        service: change.service,
+                        progress: progress,
+                        zonename: ins.zonename,
+                        log: opts.log,
+                        server: ins.server,
+                        userScript: arg.userScript
+                    }, next);
+                });
+            });
+        } else if (change.inst) {
             /*
              * If the service params require a delegate dataset, then ensure
              * the instance has one.
@@ -109,43 +154,124 @@ UpdateStatelessServicesV1.prototype.execute = function ussv1Execute(opts, cb) {
                 assert.optionalBool(change.service.params.delegate_dataset,
                     'change.service.params.delegate_dataset');
                 if (change.service.params.delegate_dataset) {
-                    steps.push(function (_, next) {
+                    funcs.push(function (_, next) {
                         s.ensureDelegateDataset({
                             service: change.service,
-                            progress: opts.progress,
-                            zonename: change.inst.zonename,
+                            progress: progress,
+                            zonename: inst.zonename,
                             log: opts.log,
-                            server: change.inst.server
+                            server: inst.server
                         }, next);
                     });
                 }
             }
 
-            steps = steps.concat([
-                s.updateVmUserScript,
-                s.imgadmInstall,
-                s.reprovision,
+            funcs.push(function (_, next) {
+                s.updateVmUserScriptRemote({
+                    service: change.service,
+                    progress: progress,
+                    zonename: inst.zonename,
+                    log: opts.log,
+                    server: inst.server,
+                    userScript: arg.userScript
+                }, next);
+            });
+        }
+
+        if (change.service.metadata) {  // workaround for assets (TOOLS-695)
+            funcs.push(s.updateSapiSvc);
+        }
+
+        if (arg.HA) {
+            change.insts.forEach(function (ins) {
+                funcs = funcs.concat(
+                    function imgadmInstallForInstance(_, next) {
+                        return s.imgadmInstallRemote({
+                            progress: progress,
+                            img: change.image,
+                            log: opts.log,
+                            server: ins.server
+                        }, next);
+                    },
+                    function reprovisionInstance(_, next) {
+                        s.reprovisionRemote({
+                            server: ins.server,
+                            img: change.image,
+                            zonename: ins.zonename,
+                            progress: progress,
+                            log: opts.log,
+                            sdcadm: opts.sdcadm
+                        }, next);
+                    },
+                    function waitForInstanceToBeUp(_, next) {
+                        s.waitForInstToBeUp({
+                            opts: {
+                                progress: progress,
+                                sdcadm: opts.sdcadm,
+                                log: opts.log
+                            },
+                            change: {
+                                inst: ins
+                            }
+                        }, next);
+                    }
+                );
+            });
+        } else if (change.inst) {
+            funcs = funcs.concat([
+                function imgadmInstall(_, next) {
+                    return s.imgadmInstallRemote({
+                        progress: progress,
+                        img: change.image,
+                        log: opts.log,
+                        server: inst.server
+                    }, next);
+                },
+                function reprovisionInst(_, next) {
+                    s.reprovisionRemote({
+                        server: inst.server,
+                        img: change.image,
+                        zonename: inst.zonename,
+                        progress: progress,
+                        log: opts.log,
+                        sdcadm: opts.sdcadm
+                    }, next);
+                },
                 s.waitForInstToBeUp
             ]);
         }
+        vasync.pipeline({funcs: funcs, arg: arg}, nextSvc);
+    }
 
-        opts.log.info({change: change}, 'UpdateStatelessServicesV1 updateSvc');
-        vasync.pipeline({funcs: steps, arg: arg}, nextSvc);
+    // TOOLS-1465: when updating individual instances, need to double check if
+    // service is part or not of an HA setup
+    function checkServiceHA(change, nextSvc) {
+        opts.sdcadm.sapi.listInstances({
+            service_uuid: change.service.uuid
+        }, function (err, insts) {
+            if (err) {
+                return nextSvc(new errors.SDCClientError(err, 'SAPI'));
+            }
+            if (insts.length > 1) {
+                change.HA = true;
+                if (!change.insts) {
+                    change.insts = [change.inst];
+                }
+            }
+            return updateService(change, nextSvc);
+        });
     }
 
-    // For now we'll update services in series.
-    // TODO: Should eventually be able to do this in parallel, or batches.
     vasync.forEachPipeline({
         inputs: self.changes,
-        func: updateSvc
+        func: checkServiceHA
     }, cb);
 };
 
 
-
 //---- exports
 
 module.exports = {
-    UpdateStatelessServicesV1: UpdateStatelessServicesV1
+    UpdateStatelessServices: UpdateStatelessServices
 };
 // vim: set softtabstop=4 shiftwidth=4:
